ui/src/shared/components/AnnotationEditorForm.tsx

```diff
@@ -0,0 +1,389 @@
+import React, {PureComponent, ChangeEvent} from 'react'
+import uuid from 'uuid'
+import moment from 'moment'
+
+import {RadioButtons, ButtonShape} from 'src/reusable_ui'
+import ConfirmButton from 'src/shared/components/ConfirmButton'
+import AnnotationTagEditorLi from 'src/shared/components/AnnotationTagEditorLi'
+
+import DefaultDebouncer, {Debouncer} from 'src/shared/utils/debouncer'
+import {BLACKLISTED_KEYS} from 'src/shared/annotations/helpers'
+
+import {Annotation} from 'src/types/annotations'
+
+const INPUT_DEBOUNCE_TIME = 600
+const DATETIME_FORMAT = 'YYYY-MM-DD HH:mm:ss.SS'
+const BAD_DATETIME_ERROR = 'Not a valid date'
+const END_BEFORE_START_ERROR = 'End date must be after start date'
+const EMPTY_TEXT_ERROR = 'Name cannot be empty'
+const DUPLICATE_KEY_ERROR = 'Tag keys must be unique'
+
+const getTime = (d: string | number): number => new Date(d).getTime()
+const isValidDate = (d: string | number): boolean => !isNaN(getTime(d))
+const formatDate = (d: string | number): string =>
+  moment(d).format(DATETIME_FORMAT)
+
+type Type = 'point' | 'window'
```

Comment:
could potentially use an enum instead of strings, especially if there's a chance of other types being added later

Reply:
Recently I've been toying with using a union type (like here) where I would normally use an enum. I think I like the union type approach, since it's more explicit and direct (can use a string for what it is is, no having to type out the enum namespace).

Reply:
another benefit of the enum to consider is that it may help counteract the fragility of string literal typos and changes

Reply:
hmm what's a scenario in which the `enum` is less fragile than a string?

The enum
```
export enum Type {
  Point = 'point',
  Window = 'window',
}
```
describes precisely the same set of values as
```
type Type = 'point' | 'window'
```
but the latter approach requires less ceremony to use (can just type `'point'` rather than `Type.Point`)